code/game/turfs/floors/natural/_natural.dm

```diff
@@ -112,6 +113,16 @@
 
 /turf/floor/natural/on_reagent_change()
 	. = ..()
+	if(reagent_type && height < 0 && reagents?.total_volume < abs(height))
+		try_refill_turf()
+
+/turf/floor/natural/proc/try_refill_turf()
+	set waitfor = FALSE
+	if(turf_is_refilling)
+		return
+	turf_is_refilling = TRUE
+	sleep(0)
+	turf_is_refilling = FALSE
```

Comment:
this will not help, it needs to be unset after the call to `add_to_reagents`.
also i think defer_update = TRUE on add_to_reagents might be better, to be honest? that would prevent the infinite recursion
if you still want to do the debounce solution what if we set turf_is_refilling to world.time and check `if(turf_refill_time >= world.time)` which would limit it to once per tick without having to sleep?